oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java

```diff
@@ -0,0 +1,455 @@
+          IdentityPoolCredentials.newBuilder((IdentityPoolCredentials) this)
+              .setServiceAccountImpersonationUrl(null)
+              .build();
+    }
+
+    String targetPrincipal = extractTargetPrincipal(serviceAccountImpersonationUrl);
+    return ImpersonatedCredentials.newBuilder()
+        .setSourceCredentials(sourceCredentials)
+        .setHttpTransportFactory(transportFactory)
+        .setTargetPrincipal(targetPrincipal)
+        .setScopes(new ArrayList<>(scopes))
+        .setLifetime(3600) // 1 hour in seconds
+        .build();
+  }
+
+  @Override
+  public Map<String, List<String>> getRequestMetadata(URI uri) throws IOException {
+    Map<String, List<String>> requestMetadata = super.getRequestMetadata(uri);
+    return addQuotaProjectIdToRequestMetadata(quotaProjectId, requestMetadata);
+  }
+
+  /**
+   * Returns credentials defined by a JSON file stream.
+   *
+   * <p>This will either return {@link IdentityPoolCredentials} or AwsCredentials.
+   *
+   * @param credentialsStream the stream with the credential definition
+   * @return the credential defined by the credentialsStream
+   * @throws IOException if the credential cannot be created from the stream
+   */
+  public static ExternalAccountCredentials fromStream(InputStream credentialsStream)
+      throws IOException {
+    return fromStream(credentialsStream, OAuth2Utils.HTTP_TRANSPORT_FACTORY);
+  }
+
+  /**
+   * Returns credentials defined by a JSON file stream.
+   *
+   * <p>This will either return a IdentityPoolCredentials or AwsCredentials.
+   *
+   * @param credentialsStream the stream with the credential definition
+   * @param transportFactory the HTTP transport factory used to create the transport to get access
+   *     tokens
+   * @return the credential defined by the credentialsStream
+   * @throws IOException if the credential cannot be created from the stream
+   */
+  public static ExternalAccountCredentials fromStream(
+      InputStream credentialsStream, HttpTransportFactory transportFactory) throws IOException {
+    checkNotNull(credentialsStream);
+    checkNotNull(transportFactory);
+
+    JsonObjectParser parser = new JsonObjectParser(OAuth2Utils.JSON_FACTORY);
+    GenericJson fileContents =
+        parser.parseAndClose(credentialsStream, StandardCharsets.UTF_8, GenericJson.class);
+    return fromJson(fileContents, transportFactory);
+  }
+
+  /**
+   * Returns external account credentials defined by JSON using the format generated by gCloud.
+   *
+   * @param json a map from the JSON representing the credentials
+   * @param transportFactory HTTP transport factory, creates the transport used to get access tokens
+   * @return the credentials defined by the JSON
+   */
+  public static ExternalAccountCredentials fromJson(
+      Map<String, Object> json, HttpTransportFactory transportFactory) {
+    checkNotNull(json);
+    checkNotNull(transportFactory);
+
+    String audience = (String) json.get("audience");
```

Comment:
There are a lot of potential class cast exceptions here, especially given that this is a public method. Does it have to be public?

Reply:
This should be package private

Reply:
can fromStream be package private too? Otherwise the same issues with class cast exceptions are just on method removed.